### DISCUSSION TITLE
Is there a way to programatically generate Markdown with directives?

### DISCUSSION THREAD

**Discussion by ofajardo** [executablebooks/meta#142] [Q&A] [answered]
Something along these lines: (backslashes introduced only to escape here in markdown)

Having this in a code cell in the notebook.
```
from IPython.display import display, Markdown

display(Markdown("""
\```{dropdown} Click on me to see my content!
I'm the content which can be **anything** {fa}`check,text-success ml-1`

:::{note}
Even other blocks.
:::
\```"""))

```

What I would like to see is the dropdown rendered correctly in the notebook. What I get now is just this:

```
I'm the content which can be **anything** {fa}`check,text-success ml-1`

:::{note}
Even other blocks.
:::
```

The idea would be to generate the dropdowns with sphinx-panels, but introduce some arbitrary content, like plots for example. 

Actually what I am aiming is to have some tabbed output with plots. I could use ipywidgets or bokeh for this, the problem I am having there is that if I have too many tabs, I cannot read the labels on them anymore as the width gets collapsed. I was hoping that maybe with sphinx panels or sphinx tabs it would look better. 
But maybe it's just doesn't work like that I guess.

**Comment by choldgraf** [comment]
you could try combining this with pasting code outputs via `glue` - see this PR I just made to mention this in the docs, does that work?

https://github.com/executablebooks/jupyter-book/pull/979

**Reply by ofajardo** [reply]
yes it works! very nice example! thanks!

**Comment by ofajardo** [comment, marked as answer]
Yep. I'll give it a try thanks. (Once I discover why my tabs are not rendering, I have another question about that)

However I was thinking more in the direction of what if I don't know in advance how many tabs are there going to be ... that's why I was thinking on generating markdown programatically. That's the way you do it in Rmarkdown.

**Reply by choldgraf** [reply]
I *believe* that you should be able to generate markdown and have it parsed with the regular parser at build time:

https://jupyterbook.org/content/code-outputs.html#markdown

try upgrading your version of jupyter book?

**Reply by chrisjsewell** [reply]
No as it says at the start of that section, markdown output is currently parsed as Commonmark only, i.e. no roles or directives. 

To allow this, whilst “guaranteeing” no issues in the parsing process you would need to devise a special (optional?) case, such that markdown outputs are parsed at the same time as the normal markdown, rather than later in the documentation build process (as is now the case in a post-transform)

**Reply by chrisjsewell** [reply]
Note this is not unfeasible, but it would require some careful consideration

**Reply by choldgraf** [reply]
ah my bad I thought that the `myst_commonmark_only` flag could be set to false in order to parse directives (https://myst-nb.readthedocs.io/en/latest/use/formatting_outputs.html#markdown) but I guess that flag only refers to non-directive extended syntax

**Reply by chrisjsewell** [reply]
No it is hard-codes to True for notebook outputs

**Reply by chrisjsewell** [reply]
The underlying reason for this is to do with the interaction between document builds and the caching process: basically sphinx generates the document trees in 1st stage, up to the latest point that it can be  “output agnostic” (e.g. not specifically HTML, TeX, etc) then saves. Then for each output format it will load this cached document tree and turn it into the specific format in “2nd stage”. With notebook outputs, they often output multiple types specialised for different output formats, so when we process the document, we save them all in the cached document tree in stage 1, then only parse them properly in this 2nd stage, selecting the correct outputs. However many roles/directives rely on manipulating the document tree in the 1st stage.

**Reply by ofajardo** [reply]
Thanks for the deep explanation!
